### PR TITLE
fix: issue 2362 and improve chain exchange failure message

### PR DESCRIPTION
**Pull request by hanabi1224**
**Summary of changes**
Changes introduced in this pull request:
- Relax chain exchange message size limit to 200MB (which is the root cause of https://github.com/ChainSafe/forest/issues/2362)
- Improve chain exchange error messages
- Add peer tipset epoch to metrics
- Reduce chain exchange batch size for message lookup to reduce bundled message size.

**Reference issue to close (if applicable)**

Closes Related to https://github.com/ChainSafe/forest/issues/2244 and https://github.com/ChainSafe/forest/issues/2362

**Other information and links**

### CODE REVIEW

File: blockchain/chain_sync/src/network_context.rs [hunk 1]
```diff
@@ -173,6 +173,8 @@ where
         };
 
         let global_pre_time = SystemTime::now();
+        let network_failures = Arc::new(RwLock::new(0));
```
Comment (hanabi1224) [inline blockchain/chain_sync/src/network_context.rs:176]: Maybe better to use Atomic here

File: blockchain/chain_sync/src/network_context.rs [hunk 2]
```diff
@@ -241,7 +248,7 @@ where
                         log::debug!("Succeed: handle_chain_exchange_request");
                         result.map_err(|e| e.to_string())?
                     },
-                    _ = wait_all(&mut tasks) => return Err("ChainExchange request failed for all top peers".into()),
+                    _ = wait_all(&mut tasks) => return Err(format!("ChainExchange request failed for all top peers. {} network failures, {} lookup failures, request:\n{request:?}", network_failures.load(Ordering::Relaxed), lookup_failures.load(Ordering::Relaxed))),
```
Comment (lemmih) [inline blockchain/chain_sync/src/network_context.rs:251]: Let's wrap this line.

File: node/forest_libp2p/src/rpc/mod.rs [hunk 3]
```diff
@@ -127,7 +127,9 @@ where
     IO: AsyncRead + Unpin,
     T: serde::de::DeserializeOwned,
 {
-    const MAX_BYTES_ALLOWED: usize = 2 * 1024 * 1024; // messages over 2MB are likely malicious
+    // FIXME: investigate the best value here, 2MB is not enough for mainnet epoch 2419928
+    // Issue: https://github.com/ChainSafe/forest/issues/2362
+    const MAX_BYTES_ALLOWED: usize = 200 * 1024 * 1024; // messages over 200MB are likely malicious
```
Comment (lemmih) [inline node/forest_libp2p/src/rpc/mod.rs:132]: 200MB messages scary me a bit. But it should be fine for now.

**Review by lemmih — APPROVED** [review]

File: blockchain/chain_sync/src/tipset_syncer.rs [hunk 4]
```diff
@@ -1076,7 +1076,7 @@ async fn sync_messages_check_state<
     invalid_block_strategy: InvalidBlockStrategy,
 ) -> Result<(), TipsetRangeSyncerError<C>> {
     // Sync the messages for one or many tipsets @ a time
-    const REQUEST_WINDOW: usize = 8;
+    const REQUEST_WINDOW: usize = 4;
```
Comment (elmattic) [inline blockchain/chain_sync/src/tipset_syncer.rs:1079]: Any reason to restrict the window to half size here?

**Review by elmattic — APPROVED** [review]

File: blockchain/chain_sync/src/tipset_syncer.rs (hunk 4 above)
Comment (elmattic) [reply blockchain/chain_sync/src/tipset_syncer.rs:1079]: This is the related optimization: https://github.com/ChainSafe/forest/pull/1943

Comment (hanabi1224) [reply blockchain/chain_sync/src/tipset_syncer.rs:1079]: >- Reduce chain exchange batch size for message lookup to reduce bundled message size.

The larger chain exchange response is, the more we try cbor-decoding it partially (with failure)

File: blockchain/chain_sync/src/network_context.rs (hunk 2 above)
Comment (hanabi1224) [reply blockchain/chain_sync/src/network_context.rs:251]: Fixed

File: blockchain/chain_sync/src/tipset_syncer.rs (hunk 4 above)
Comment (elmattic) [reply blockchain/chain_sync/src/tipset_syncer.rs:1079]: Ah ok, get it now 👍

**Pull request merged**